cpp/src/stream_compaction/stream_compaction.cu

```diff
@@ -108,4 +115,118 @@ gdf_column drop_nulls(gdf_column const &input) {
   }
 }
 
+
+namespace detail {
+struct tuple_reduced_and
+{
+    __host__ __device__ bool operator()(const thrust::tuple<bool,bool> &x) const {
+        return thrust::get<0>(x) && thrust::get<1>(x); }
+}; 
+
+template<typename DerivedPolicy,
+         typename InputIterator,
+         typename OutputIterator,
+         typename BinaryPredicate>
+__host__ __device__
+  OutputIterator non_duplicates_copy(thrust::execution_policy<DerivedPolicy> &exec,
+                             InputIterator first,
+                             InputIterator last,
+                             OutputIterator output,
+                             BinaryPredicate binary_pred)
+{
+  thrust::detail::head_flags<InputIterator, BinaryPredicate> stencil1(first, last, binary_pred);
+  thrust::detail::tail_flags<InputIterator, BinaryPredicate> stencil2(first, last, binary_pred);
+  auto combined_stencil = thrust::make_transform_iterator(
+          thrust::make_zip_iterator(thrust::make_tuple(stencil1.begin(), stencil2.begin())),
+          tuple_reduced_and());
+  
+  using namespace thrust::placeholders;
+  
+  return thrust::copy_if(exec, first, last, combined_stencil, output, _1);
+} 
+
+rmm::device_vector<gdf_index_type>
+get_unique_ordered_indices(const cudf::table& key_columns,
+                           const duplicate_keep_option keep)
+{
+  gdf_size_type ncols = key_columns.num_columns();
+  gdf_size_type nrows = key_columns.num_rows();
+
+  // sort only indices
+  rmm::device_vector<gdf_size_type> sorted_indices(nrows);
+  gdf_context context;
+  gdf_column sorted_indices_col;
+  CUDF_TRY(gdf_column_view(&sorted_indices_col, (void*)(sorted_indices.data().get()),
+        nullptr, nrows, GDF_INT32));
+  CUDF_TRY(gdf_order_by(key_columns.begin(),
+        nullptr,
+        key_columns.num_columns(),
+        &sorted_indices_col,
+        &context));
+
+  // extract unique indices 
+  rmm::device_vector<gdf_index_type> unique_indices(nrows);
+
+  cudaStream_t stream;
+  cudaStreamCreate(&stream);
+  auto exec = rmm::exec_policy(stream)->on(stream);
+
+  auto device_input_table = device_table::create(key_columns, stream);
+  auto comp = row_equality_comparator<true>(*device_input_table, true);
```

Comment:
As an optimization, you should test if the input table actually has nulls, and then select `row_equality_comparator<true/false>` accordingly.

Reply:
Implemented. 
Note: Compile time increased.